Further relax VariableFlags, ... and fix bugs

zdevito opened the pull request:
* Allow a requires_grad=True trace to be used for a requires_grad=False
  input by computing the gradient but they not connecting it to the
  input.
* Enable CSE to de-duplicate WLM backwards pass code which calls sum twice.
* Fix a bug in the interpreter that frees a register too early when
  it appears twice in a use list.

zdevito committed:
Further relax VariableFlags

* Allow a requires_grad=True trace to be used for a requires_grad=False
  input by computing the gradient but they not connecting it to the
  input.
* Enable CSE to de-duplicate WLM backwards pass code which calls sum twice.
* Fix a bug in the interpreter that frees a register too early when
  it appears twice in a use list.

ezyang commented on torch/csrc/jit/variable_flags.h:
```diff
@@ -8,16 +8,15 @@ namespace torch { namespace jit {
 
   bool requires_grad;
-  bool was_null;
+  bool defined;
```
Good.

ezyang commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -289,12 +289,17 @@ struct CodeImpl {
     auto scanUses = [&](UseList & u) {
-      listBegin(u.free_flags);
-      for(int i = 0; i < u.values.size; i++) {
+      // scan backwards because the same value may appear > once in a use list
```
Is this bugfix exercised by the tests now?

zdevito replied:
no -- I am going to do a separate PR that has a concise test harness for JIT so we can easily test corner cases.

apaszke replied:
I don't fully understand this fix. This function scans an input list to every instruction in reverse, so it doesn't really matter which input marks the value to be freed later, right? Or do we empty the registers when we load inputs and that's what's causing the failure (next inputs get undefined tensors)? A bit more detailed comment would be helpful

ezyang commented:
LGTM.

zdevito committed:
[fuser] Follow all outputs to check if fusion is safe

This bug was introduced when we allowed fusion groups
to fuse together. Previously producers were forced to have a single
output, but now producers that are fusion groups can have multiple outputs.
So now we check the uses of all the outputs of a producer.

zdevito committed:
[JIT] Fix handling of undefined inputs

It is not legal to call .data() on variable objects whose tensors
are undefined.

zdevito commented:
I am appending more simple bug fixes here, so let's wait to merge. 

I find it very weird that a Variable can be undefined, and when it is, .data() just segfaults. This makes me concerned that there are a lot of other places that don't handle undefined well.

apaszke reviewed (approved):
LGTM. Would be nice to clarify that comment. Also, would be good to add more tests

ezyang commented:
Re undefined not being handled correctly, see also https://github.com/pytorch/pytorch/issues/4215

zdevito commented:
Yes, the backward scan is because the register -> outputs pathway frees things from the register list as it goes. I'll clarify. I am going to put tests in another PR because I want to get this merged and get the alpha documents out first. The test commit will be about testing all the corner cases: no-op graphs, use inputs 0, 1, 2 times, maybe attempt to create every graph + backward of a single instruction to make sure the trace continues to work.
